cirq/qis/states.py

```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
+    if there is a unique compatible qid shape. If, after applying the heuristic,
+    the qid shape is ambiguous, then an error is raised unless the qid shape is
+    to be inferred from the shape of a state vector or density matrix. In that
+    case, if the dimension of the state space is a power of 2, then the state
+    space is assumed to be composed of qubits; otherwise, it is assumed to be
+    composed of a single qudit.
+
+    Args:
+        states: The states for which to infer the qid shape.
+
+    Returns:
+        The inferred qid shape.
+
+    Raises:
+        ValueError: The qid shape of the given states is ambiguous.
+        ValueError: Failed to infer the qid shape of the given states.
+    """
+    integer_states = []
+    non_integer_states = []
+    for state in states:
+        if isinstance(state, int):
+            integer_states.append(state)
+        else:
+            non_integer_states.append(state)
+
+    # if all states are specified as integers, the shape is ambiguous
+    if not non_integer_states:
+        raise ValueError('The qid shape of the given states is ambiguous.')
+
+    # attempt to infer qid shape from non-integer states
+    potential_shapes = _potential_qid_shapes(non_integer_states[0])
+    for state in non_integer_states[1:]:
+        potential_shapes = potential_shapes.intersection_subset(_potential_qid_shapes(state))
+    if potential_shapes.is_ambiguous():
+        raise ValueError('The qid shape of the given states is ambiguous.')
+    qid_shape = potential_shapes.infer_qid_shape()
+    if qid_shape is None:
+        raise ValueError('Failed to infer the qid shape of the given states.')
+
+    # check if the shape is compatible with the states specified as integers
+    if integer_states and np.prod(qid_shape, dtype=int) <= max(integer_states):
+        raise ValueError('Failed to infer the qid shape of the given states.')
+
+    return qid_shape
+
+
+_NON_INT_STATE_LIKE = Union[
+    # Per-qudit computational basis values.
+    Sequence[int],
+    # Explicit state vector or state tensor.
+    np.ndarray,
+    Sequence[Union[int, float, complex]],
+    # Product state object
+    'cirq.ProductState',
+    # Quantum state object
+    'cirq.QuantumState',
+]
+
+
+def _potential_qid_shapes(state: _NON_INT_STATE_LIKE) -> '_QidShapeSet':
+    if isinstance(state, QuantumState):
+        return _QidShapeSet(explicit_qid_shapes={state.qid_shape})
+    if isinstance(state, value.ProductState):
+        return _QidShapeSet(explicit_qid_shapes={(2,) * len(state)})
+
+    if isinstance(state, Sequence):
+        state = np.array(state)
+    if state.ndim == 1:
+        (dim,) = state.shape
+        if state.dtype.kind == 'c':
+            # definitely state vector amplitudes
+            return _QidShapeSet(unfactorized_total_dimension=dim)
+        # could also be per-qudit computational basis values
+        min_qudit_dimensions = tuple(state.astype(int, copy=False) + 1)
```

Comment:
what if `state.dtype.kind` is float (i.e. it just happens to not have complex elements)

Reply:
I've changed the code to check the type before converting to int.